### PR TITLE
feature(batch-action) add form provider

### PR DESCRIPTION
**Pull request by wfajczyk**
# Description
Add form provider 

Issue # 
## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

- [ ] Unit Tests
- [ ] Behat Test

# Checklist:
- [ ] I have read the contribution requirements and fulfill them.
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: module/batch-action/src/Application/Controller/Api/CreateBatchAction.php [hunk 1]
```diff
@@ -73,8 +78,10 @@ public function __construct(
      */
     public function __invoke(Request $request): Response
     {
+        $type = $request->request->get('type');
         try {
-            $form = $this->formFactory->create(BatchActionForm::class);
+            $class = $this->formProvider->provide($type);
+            $form = $this->formFactory->create($class ?: BatchActionForm::class);
```
Comment (wfajczyk) [inline module/batch-action/src/Application/Controller/Api/CreateBatchAction.php:84]: Take the default form if there is no custom one.

File: module/product/src/Application/Form/Product/BatchAction/ProductEditForm.php [hunk 2]
```diff
@@ -6,44 +6,61 @@
...
+use Ergonode\BatchAction\Application\Form\Type\BatchActionFilterType;
+use Ergonode\Product\Application\Form\Product\Attribute\Update\UpdateAttributeValueForm;
 use Symfony\Component\Form\AbstractType;
 use Symfony\Component\Form\Extension\Core\Type\CollectionType;
 use Symfony\Component\Form\Extension\Core\Type\TextType;
 use Symfony\Component\Form\FormBuilderInterface;
 use Symfony\Component\OptionsResolver\OptionsResolver;
 
-class BatchActionPayloadType extends AbstractType
+class ProductEditForm extends AbstractType implements BatchActionFormInterface
```
Comment (wfajczyk) [inline module/product/src/Application/Form/Product/BatchAction/ProductEditForm.php:21]: Type 'product_edit'
Used
`ProductChangeAttributeBatchActionProcessor`
`ChangeProductAttributesCommand`
I thing is beter type name:
`product_attribute_change `

File: module/batch-action/src/Application/Provider/BatchActionFormProvider.php [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    public function provide(string $type): ?string
+    {
+        foreach ($this->forms as $form) {
+            if ($form->supported($type)) {
+                return get_class($form);
+            }
+        }
+
+        return null;
+        throw new \RuntimeException(sprintf('Can\' find factory for %s type', $type));
```
Comment (rprzedzik) [inline module/batch-action/src/Application/Provider/BatchActionFormProvider.php:40]: This exception will never be thrown

File: module/batch-action/src/Application/Provider/BatchActionFormProvider.php [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+     * @param BatchActionFormInterface[] $forms
+     */
+    public function __construct(iterable $forms)
+    {
+        Assert::allIsInstanceOf($forms, BatchActionFormInterface::class);
+
+        $this->forms = $forms;
+    }
+
+    public function provide(string $type): ?string
```
Comment (rprzedzik) [inline module/batch-action/src/Application/Provider/BatchActionFormProvider.php:31]: IMHO its be better move default Form Type hire and always return form class insisted of return null

File: module/batch-action/src/Infrastructure/Provider/BatchActionProcessorInterface.php [hunk 5]
```diff
@@ -19,11 +18,13 @@ interface BatchActionProcessorInterface
     public function supports(BatchActionType $type): bool;
 
     /**
+     * @param mixed $payload
+     *
```
Comment (wiewiurdp) [inline module/batch-action/src/Infrastructure/Provider/BatchActionProcessorInterface.php:22]: do we need this annotation ?

**Review by wiewiurdp — APPROVED** [review]

**Review by rprzedzik — APPROVED** [review]

File: module/batch-action/src/Infrastructure/Provider/BatchActionProcessorInterface.php (hunk 5 above)
Comment (wfajczyk) [reply module/batch-action/src/Infrastructure/Provider/BatchActionProcessorInterface.php:22]: yes

**Pull request merged**